File: python/src/openharness/tracing.py
"""Tracing: see every model call, tool call, handoff and guardrail in a run.

Each run creates a ``Trace`` made of nested ``Span``s. Finished traces go to
processors. Built-in processors print to the console or append JSON Lines to
a file; ``OpenTelemetryProcessor`` forwards spans to any OTel backend when
``opentelemetry-api`` is installed.

    run(agent, "hi", trace=ConsoleProcessor())
    # or globally:
    add_trace_processor(JSONLProcessor("traces.jsonl"))
    # or with no code change:
    OPENHARNESS_TRACE=console | OPENHARNESS_TRACE=./traces.jsonl
"""

from __future__ import annotations

import contextlib
import json
import os
import sys
import threading
import time
from collections.abc import Iterator
from dataclasses import dataclass, field
from pathlib import Path
from typing import Any, Protocol, TextIO

from .types import new_id


@dataclass
class Span:
    kind: str  # run | agent | model | tool | guardrail | handoff | mcp
    name: str
    trace_id: str
    id: str = field(default_factory=lambda: new_id("span"))
    parent_id: str | None = None
    start: float = field(default_factory=time.time)
    end: float | None = None
    attributes: dict[str, Any] = field(default_factory=dict)
    error: str | None = None

    @property
    def duration_ms(self) -> float | None:
        return None if self.end is None else round((self.end - self.start) * 1000, 1)

    def set(self, **attrs: Any) -> None:
        self.attributes.update(attrs)

    def to_dict(self) -> dict[str, Any]:
        return {
            "id": self.id,
            "trace_id": self.trace_id,
            "parent_id": self.parent_id,
            "kind": self.kind,
            "name": self.name,
            "start": self.start,
            "end": self.end,
            "duration_ms": self.duration_ms,
            "attributes": self.attributes,
            "error": self.error,
        }


class TraceProcessor(Protocol):
    def on_span_start(self, span: Span) -> None: ...
    def on_span_end(self, span: Span) -> None: ...
    def on_trace_end(self, trace: Trace) -> None: ...


class Trace:
    def __init__(
        self, name: str, processors: list[TraceProcessor] | None = None, metadata: dict[str, Any] | None = None
    ):
        self.id = new_id("trace")
        self.name = name
        self.metadata = metadata or {}
        self.spans: list[Span] = []
        self.processors = processors or []
        self._stack: list[Span] = []

    @contextlib.contextmanager
    def span(self, kind: str, name: str, parent: Span | None = None, **attrs: Any) -> Iterator[Span]:
        """Open a span. Pass ``parent`` for work that runs concurrently (it is then kept off the stack)."""
        concurrent = parent is not None
        parent_id = parent.id if parent else (self._stack[-1].id if self._stack else None)
        sp = Span(kind=kind, name=name, trace_id=self.id, parent_id=parent_id, attributes=dict(attrs))
        self.spans.append(sp)
        if not concurrent:
            self._stack.append(sp)
        self._emit("on_span_start", sp)
        try:
            yield sp
        except BaseException as e:
            sp.error = f"{type(e).__name__}: {e}"
            raise
        finally:
            sp.end = time.time()
            if sp in self._stack:
                self._stack.remove(sp)
            self._emit("on_span_end", sp)

    def finish(self) -> None:
        self._emit("on_trace_end", self)

    def _emit(self, method: str, arg: Any) -> None:
        for p in self.processors:
            try:
                getattr(p, method)(arg)
            except Exception as e:  # never let tracing break a run
                print(f"[openharness] trace processor {type(p).__name__} failed: {e}", file=sys.stderr)

    def to_dict(self) -> dict[str, Any]:
        return {"id": self.id, "name": self.name, "metadata": self.metadata, "spans": [s.to_dict() for s in self.spans]}


# ------------------------------------------------------------------ processors


class ConsoleProcessor:
    """Prints one line per finished span, indented by depth."""

    def __init__(self, stream: TextIO | None = None, show_attributes: bool = False):
        self.stream = stream or sys.stderr
        self.show_attributes = show_attributes
        self._depth: dict[str, int] = {}

    def on_span_start(self, span: Span) -> None:
        self._depth[span.id] = self._depth.get(span.parent_id or "", -1) + 1

    def on_span_end(self, span: Span) -> None:
        pad = "  " * self._depth.pop(span.id, 0)
        status = f" ERROR {span.error}" if span.error else ""
        extra = ""
        a = span.attributes
        if span.kind == "model" and "usage" in a:
            u = a["usage"]
            extra = f" in={u.get('input_tokens')} out={u.get('output_tokens')}"
        if self.show_attributes and a:
            extra += " " + json.dumps(a, default=str)[:300]
        print(f"[trace] {pad}{span.kind}:{span.name} {span.duration_ms}ms{extra}{status}", file=self.stream)

    def on_trace_end(self, trace: Trace) -> None:
        return None


class JSONLProcessor:
    """Appends each finished trace as one JSON line."""

    def __init__(self, path: str | Path):
        self.path = Path(path).expanduser()
        self._lock = threading.Lock()

    def on_span_start(self, span: Span) -> None:
        return None

    def on_span_end(self, span: Span) -> None:
        return None

    def on_trace_end(self, trace: Trace) -> None:
        with self._lock:
            self.path.parent.mkdir(parents=True, exist_ok=True)
            with self.path.open("a", encoding="utf-8") as f:
                f.write(json.dumps(trace.to_dict(), default=str) + "\n")


class MemoryProcessor:
    """Keeps finished traces in a list. Handy for tests and dashboards."""

    def __init__(self) -> None:
        self.traces: list[Trace] = []

    def on_span_start(self, span: Span) -> None:
        return None

    def on_span_end(self, span: Span) -> None:
        return None

    def on_trace_end(self, trace: Trace) -> None:
        self.traces.append(trace)


class OpenTelemetryProcessor:
    """Re-emits spans through OpenTelemetry. Requires ``opentelemetry-api``."""

    def __init__(self, tracer_name: str = "openharness"):
        from opentelemetry import trace as otel  # type: ignore[import-not-found]

        self._otel = otel
        self._tracer = otel.get_tracer(tracer_name)
        self._live: dict[str, Any] = {}

    def on_span_start(self, span: Span) -> None:
        parent = self._live.get(span.parent_id or "")
        ctx = self._otel.set_span_in_context(parent) if parent is not None else None
        self._live[span.id] = self._tracer.start_span(
            f"{span.kind} {span.name}", context=ctx, start_time=int(span.start * 1e9)
        )

    def on_span_end(self, span: Span) -> None:
        s = self._live.pop(span.id, None)
        if s is None:
            return
        for k, v in span.attributes.items():
            s.set_attribute(
                f"openharness.{k}", v if isinstance(v, (str, int, float, bool)) else json.dumps(v, default=str)
            )
        if span.error:
            s.set_attribute("error", span.error)
        s.end(end_time=int((span.end or time.time()) * 1e9))

    def on_trace_end(self, trace: Trace) -> None:
        return None


_GLOBAL: list[TraceProcessor] = []


def add_trace_processor(processor: TraceProcessor) -> None:
    _GLOBAL.append(processor)


def clear_trace_processors() -> None:
    _GLOBAL.clear()


def processors_from_env() -> list[TraceProcessor]:
    value = os.environ.get("OPENHARNESS_TRACE", "").strip()
    if not value or value.lower() in ("0", "false", "off", "none"):
        return []
    if value.lower() in ("1", "true", "console", "stderr"):
        return [ConsoleProcessor()]
    return [JSONLProcessor(value)]


def global_processors() -> list[TraceProcessor]:
    return [*_GLOBAL, *processors_from_env()]

File: typescript/src/tracing.ts
/**
 * Tracing: see every model call, tool call, handoff and guardrail in a run.
 *
 * Each run creates a `Trace` of nested `Span`s. Finished traces go to
 * processors: print to the console, append JSON Lines, or forward anywhere.
 *
 *   run(agent, "hi", { trace: new ConsoleProcessor() })
 *   addTraceProcessor(new JSONLProcessor("traces.jsonl"))
 *   // or with no code change: OPENHARNESS_TRACE=console | OPENHARNESS_TRACE=./traces.jsonl
 */

import { appendFileSync, mkdirSync } from "node:fs";
import { dirname } from "node:path";
import { newId } from "./types.js";

export class Span {
  readonly id = newId("span");
  readonly start = Date.now() / 1000;
  end?: number;
  error?: string;
  constructor(
    readonly kind: string,
    readonly name: string,
    readonly traceId: string,
    readonly parentId: string | undefined,
    readonly attributes: Record<string, unknown> = {},
  ) {}

  get durationMs(): number | undefined {
    return this.end === undefined ? undefined : Math.round((this.end - this.start) * 10000) / 10;
  }

  set(attrs: Record<string, unknown>): void {
    Object.assign(this.attributes, attrs);
  }

  toJSON() {
    return {
      id: this.id,
      trace_id: this.traceId,
      parent_id: this.parentId ?? null,
      kind: this.kind,
      name: this.name,
      start: this.start,
      end: this.end ?? null,
      duration_ms: this.durationMs ?? null,
      attributes: this.attributes,
      error: this.error ?? null,
    };
  }
}

export interface TraceProcessor {
  onSpanStart?(span: Span): void;
  onSpanEnd?(span: Span): void;
  onTraceEnd?(trace: Trace): void;
}

export class Trace {
  readonly id = newId("trace");
  readonly spans: Span[] = [];

  constructor(
    readonly name: string,
    readonly processors: TraceProcessor[] = [],
    readonly metadata: Record<string, unknown> = {},
  ) {}

  startSpan(kind: string, name: string, parent?: Span, attributes: Record<string, unknown> = {}): Span {
    const span = new Span(kind, name, this.id, parent?.id, { ...attributes });
    this.spans.push(span);
    this.emit((p) => p.onSpanStart?.(span));
    return span;
  }

  endSpan(span: Span, error?: unknown): void {
    if (span.end !== undefined) return;
    span.end = Date.now() / 1000;
    if (error) span.error = error instanceof Error ? `${error.name}: ${error.message}` : String(error);
    this.emit((p) => p.onSpanEnd?.(span));
  }

  /** Run `fn` inside a span, ending it (and recording any error) afterwards. */
  async span<T>(kind: string, name: string, parent: Span | undefined, attrs: Record<string, unknown>, fn: (span: Span) => Promise<T>): Promise<T> {
    const span = this.startSpan(kind, name, parent, attrs);
    try {
      const out = await fn(span);
      this.endSpan(span);
      return out;
    } catch (e) {
      this.endSpan(span, e);
      throw e;
    }
  }

  finish(): void {
    this.emit((p) => p.onTraceEnd?.(this));
  }

  private emit(fn: (p: TraceProcessor) => void): void {
    for (const p of this.processors) {
      try {
        fn(p);
      } catch (e) {
        console.error(`[openharness] trace processor failed: ${(e as Error).message}`);
      }
    }
  }

  toJSON() {
    return { id: this.id, name: this.name, metadata: this.metadata, spans: this.spans.map((s) => s.toJSON()) };
  }
}

/** Prints one line per finished span, indented by depth. */
export class ConsoleProcessor implements TraceProcessor {
  private depth = new Map<string, number>();
  constructor(private readonly write: (line: string) => void = (l) => process.stderr.write(l + "\n")) {}

  onSpanStart(span: Span): void {
    this.depth.set(span.id, (span.parentId ? (this.depth.get(span.parentId) ?? -1) : -1) + 1);
  }

  onSpanEnd(span: Span): void {
    const pad = "  ".repeat(this.depth.get(span.id) ?? 0);
    let extra = "";
    const u = span.attributes.usage as any;
    if (span.kind === "model" && u) extra = ` in=${u.inputTokens} out=${u.outputTokens}`;
    this.write(`[trace] ${pad}${span.kind}:${span.name} ${span.durationMs}ms${extra}${span.error ? ` ERROR ${span.error}` : ""}`);
  }
}

/** Appends each finished trace as one JSON line. */
export class JSONLProcessor implements TraceProcessor {
  constructor(readonly path: string) {}
  onTraceEnd(trace: Trace): void {
    mkdirSync(dirname(this.path), { recursive: true });
    appendFileSync(this.path, JSON.stringify(trace) + "\n", "utf8");
  }
}

/** Keeps finished traces in memory. Handy for tests and dashboards. */
export class MemoryProcessor implements TraceProcessor {
  readonly traces: Trace[] = [];
  onTraceEnd(trace: Trace): void {
    this.traces.push(trace);
  }
}

const GLOBAL: TraceProcessor[] = [];
export const addTraceProcessor = (p: TraceProcessor): void => void GLOBAL.push(p);
export const clearTraceProcessors = (): void => void GLOBAL.splice(0);

export function processorsFromEnv(): TraceProcessor[] {
  const v = (process.env.OPENHARNESS_TRACE ?? "").trim();
  if (!v || ["0", "false", "off", "none"].includes(v.toLowerCase())) return [];
  if (["1", "true", "console", "stderr"].includes(v.toLowerCase())) return [new ConsoleProcessor()];
  return [new JSONLProcessor(v)];
}

export function globalProcessors(): TraceProcessor[] {
  return [...GLOBAL, ...processorsFromEnv()];
}
